maitake/src/sync/mutex.rs

```diff
@@ -0,0 +1,487 @@
+//!
+//! See the documentation on the [`Mutex`] type for details.
+use crate::{
+    loom::cell::{MutPtr, UnsafeCell},
+    wait::queue::{self, WaitQueue},
+};
+use core::{
+    future::Future,
+    ops::{Deref, DerefMut},
+    pin::Pin,
+    task::{Context, Poll},
+};
+use mycelium_util::{fmt, unreachable_unchecked};
+use pin_project::pin_project;
+
+/// An asynchronous [mutual exclusion lock][mutex] for protecting shared data.
+///
+/// The data can only be accessed through the [RAII guards] returned
+/// from [`lock`] and [`try_lock`], which guarantees that the data is only ever
+/// accessed when the mutex is locked.
+///
+/// # Comparison With Other Mutices
+///
+/// This is an *asynchronous* mutex. When the shared data is locked, the
+/// [`lock`] method will wait by causing the current [task] to yield until the
+/// shared data is available. This is in contrast to *blocking* mutices, such as
+/// [`std::sync::Mutex`], which wait by blocking the current thread[^1], or
+/// *spinlock* based mutices, such as [`mycelium_util::sync::spin::Mutex`],
+/// which wait by spinning in a busy loop.
+///
+/// The [`futures-util`] crate also provides an implementation of an asynchronous
+/// mutex, [`futures_util::lock::Mutex`]. However, this mutex requires the Rust
+/// standard library, and is thus unsuitable for use in environments where the
+/// standard library is unavailable. In addition, the `futures-util` mutex
+/// requires an additional allocation for every task that is waiting to acquire
+/// the lock, while `maitake`'s mutex is based on an [intrusive linked list],
+/// and therefore can be used without allocation[^2]. This makes `maitake`'s
+/// mutex suitable for environments where heap allocations must be minimized or
+/// cannot be used at all.
+///
+/// In addition, this is a [fairly queued] mutex. This means that the lock is
+/// always acquired in a first-in, first-out order &mdash; if a task acquires
+/// and then releases the lock, and then wishes to acquire the lock again, it
+/// will not acquire the lock until every other task ahead of it in the queue
+/// has had a chance to lock the shared data. Again, this is in contrast to
+/// [`std::sync::Mutex`], where fairness depends on the underlying OS' locking
+/// primitives; and [`mycelium_util::sync::spin::Mutex`] and
+/// [`futures_util::lock::Mutex`], which will never guarantee fairness.
+///
+/// Finally, this mutex does not implement [poisoning][^3], unlike
+/// [`std::sync::Mutex`].
+///
+/// [^1]: And therefore require an operating system to manage threading.
+///
+/// [^2]: The [tasks][crate::task::Task] themselves must, of course, be stored
+///     somewhere, but this need not be a heap allocation in systems with a
+///     fixed set of  statically-allocated tasks. And, when tasks *are*
+///     heap-allocated, these  allocations [need not be provided by
+///     `liballoc`][storage].
+///
+/// [^3]: In fact, this mutex _cannot_ implement poisoning, as poisoning
```

Comment:
Is this documented anywhere else?

Reply:
no, i keep meaning to add it to the crate level docs

Reply:
Not a blocker! I just wanted to cross link it if there was a larger discussion somewhere.